From bb99eafe763bae8424b5302f03bbf48ea256f5f4 Mon Sep 17 00:00:00 2001
From: Jianxiong Pan <jianxiong.pan@amlogic.com>
Date: Tue, 30 Aug 2022 09:39:57 +0800
Subject: [PATCH 40/95] mm: executable code not into cma pool. [1/1]

PD#SWPL-92663

Problem:
avoid the text code into cma pool, affect cma allocation.

Solution:
exclude executable code into cma pool.

Verify:
sc2_ah212.

Change-Id: I3a17b1bfb5ab416f543e40bdbbca617cf70c02c5
Signed-off-by: Jianxiong Pan <jianxiong.pan@amlogic.com>
---
 mm/memory.c | 6 ++++++
 1 file changed, 6 insertions(+)

diff --git a/mm/memory.c b/mm/memory.c
index e111010e90d9..26ee674bb896 100644
--- a/mm/memory.c
+++ b/mm/memory.c
@@ -4864,6 +4864,12 @@ static vm_fault_t __handle_mm_fault(struct vm_area_struct *vma,
 	p4d_t *p4d;
 	vm_fault_t ret;
 
+#ifdef CONFIG_AMLOGIC_CMA
+	if (vma->vm_file && vma->vm_file->f_mapping &&
+	    (vma->vm_flags & VM_EXEC))
+		vma->vm_file->f_mapping->gfp_mask |= __GFP_NO_CMA;
+#endif
+
 #ifdef CONFIG_SPECULATIVE_PAGE_FAULT
 	if (flags & FAULT_FLAG_SPECULATIVE) {
 		pgd_t pgdval;
-- 
2.25.1

